skrub/_datetime_encoder.py

```diff
@@ -18,62 +21,494 @@
+
+    for col_idx in indices:
+        X_split[col_idx] = pd.to_datetime(
+            X_split[col_idx], format=indice_to_format[col_idx], **kwargs
+        )
+
+    return X_split
+
+
+def _get_datetime_column_indices(X_split):
+    """Select the datetime parsable columns by their indices \
+    and return their datetime format.
+
+    Parameters
+    ----------
+    X_split : list of 1d array of length n_features
+
+    Returns
+    -------
+    datetime_indices : list of int
+        List of parsable column, identified by their indices.
+
+    indice_to_format: mapping of int to str
+        Dictionary mapping parsable column indices to their datetime format.
+    """
+    indices = []
+    indice_to_format = {}
+
+    for col_idx, X_col in enumerate(X_split):
+        X_col = X_col[pd.notnull(X_col)]
+        if _is_column_datetime_parsable(X_col):
+            indices.append(col_idx)
+            # TODO: pass require_dayfirst to _guess_datetime_format
+            indice_to_format[col_idx] = _guess_datetime_format(X_col)
+
+    return indices, indice_to_format
+
+
+def _is_column_datetime_parsable(X_col):
+    """Check whether a 1d array can be converted into a \
+    :class:`pandas.DatetimeIndex`.
+
+    Parameters
+    ----------
+    X_col : array-like of shape ``(n_samples,)``
+
+    Returns
+    -------
+    is_dt_parsable : bool
+    """
+    # Remove columns of int, float or bool casted as object.
+    try:
+        if np.array_equal(X_col, X_col.astype(np.float64)):
+            return False
+    except ValueError:
+        pass
+
+    np_dtypes_candidates = [np.object_, np.str_, np.datetime64]
+    is_type_datetime_compatible = any(
+        np.issubdtype(X_col.dtype, np_dtype) for np_dtype in np_dtypes_candidates
+    )
+    if is_type_datetime_compatible:
+        try:
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore", category=UserWarning)
+                # format=mixed parses entries individually,
+                # avoiding ValueError when both date and datetime formats
+                # are present.
+                # At this stage, the format itself doesn't matter.
+                _ = pd.to_datetime(X_col, format="mixed")
+            return True
+        except (pd.errors.ParserError, ValueError):
+            pass
+    return False
+
+
+def _guess_datetime_format(X_col, require_dayfirst=True):
+    """
+    Parameters
+    ----------
+    X_col : ndarray of shape ``(n_samples,)``
+
+    require_dayfirst : bool, default True
+        Whether to return the dayfirst format when both dayfirst
+        and monthfirst are valid.
+
+    Returns
+    -------
+    format : str
+    """
+    X_col = X_col.astype(np.object_)
+    vfunc = np.vectorize(guess_datetime_format)
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", category=UserWarning)
+        month_first_formats = np.unique(vfunc(X_col, dayfirst=False))
```

Comment:
this will raise a TypeError on datetime inputs:
```python
import pandas as pd
from skrub import to_datetime

df = pd.DataFrame({"date": ["2023-10-12T17:36:50"]})
df["date"] = pd.to_datetime(df["date"])

to_datetime(df)
```

(Note this is a different error than the one  in `is_datetime_parseable` mentioned above)